server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -335,11 +340,53 @@ private static void sendErrorResponse(String actionName, TransportChannel transp
     private <T extends TransportResponse> void handleResponse(
         InetSocketAddress remoteAddress,
         final StreamInput stream,
-        final TransportResponseHandler<T> handler
+        final TransportResponseHandler<T> handler,
+        final InboundMessage inboundMessage
+    ) {
+        final String executor = handler.executor();
+        if (ThreadPool.Names.SAME.equals(executor)) {
+            // no need to provide a buffer release here, we never escape the buffer when handling directly
+            doHandleResponse(handler, remoteAddress, stream, inboundMessage.getHeader(), () -> {});
+        } else {
+            inboundMessage.incRef();
+            // release buffer once we deserialize the message, but have a fail-safe in #onAfter below in case that didn't work out
+            final Releasable releaseBuffer = Releasables.releaseOnce(inboundMessage::decRef);
+            threadPool.executor(executor).execute(new ForkingResponseHandlerRunnable(handler, null) {
+                @Override
+                protected void doRun() {
+                    doHandleResponse(handler, remoteAddress, stream, inboundMessage.getHeader(), releaseBuffer);
```

Comment:
I guess I don't totally understand why we need to pass the `releaseBuffer` mechanism into the method here. 

`onAfter` already handles the release. I'm not totally clear why it matters if the `doHandleResponse` method clearly releases the thing. It's already being released in `onAfter` no matter what.

Reply:
The motivation here was to release the buffer asap and not needlessly hold on to it until the handler is is done with the deserialized message. The `onAfter` was just put in place as a final fail-safe.

Reply:
In fact we should always release it in `doHandleResponse`, a response handler should never be rejected (see assertions in `ForkingResponseHandlerRunnable`) and there's no chance an exception could prevent it either.

That said, I'm :+1: on paranoid leak prevention.